### PR TITLE
Pydantic configs

### PR DESCRIPTION
**Pull request by arik-shurygin**
trying to settle on a schema for Pydantic configs capable to representing a wider array of Compartmental models as well as being more explicit than the previous config style.

Proposed Structure goes as follows:

lowest classes are `Bins` and `Strain`. A `Bin` such as the `DiscretizedPositiveIntBin` is usually a part of a `Dimension`, which is in turn a part of a `Compartment`.

Thus you might have something like:
```
# simple class name wrapper for interpretability
class AgeBin(DiscretizedPositiveIntBin):
    pass

age_dimension = Dimension(
            name="age",
            bins=[
                AgeBin(min_value=0, max_value=17),
                AgeBin(min_value=18, max_value=49),
                AgeBin(min_value=50, max_value=64),
                AgeBin(min_value=65, max_value=99),
            ],
        )

s_compartment = Compartment(
            name="s",
            dimensions=[
                age_dimension,
                ...
            ],
        )
```
Similarly you could build up your vaccination, infection history, and waning dimensions similarly. We will have support for a number of different types of bins as well as helper methods for creation of fully stratified or single strain immune history dimensions.

One level above this is a `CompartmentModel` which simply defines a `list[Compartment]` as well as an `ode_function` which returns the gradients of those compartments.

Finally we have a `ParamStore` to hold information like the number of MCMC warmups/samples and other information not directly linked to solving ODEs.

see `src/dynode/config/covid_seip_config.py` for an example of how an existing experiment may be reformatted to the new config style.

### CODE REVIEW

File: src/dynode/config/covid_seip_config.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+
+class SEIPModel(CompartmentalModel):
+    def __init__(self):
+        strains = self._get_strains()
+        compartments = self._get_compartments(strains)
+        param_store = self._get_param_store()
+        # Here you pass in a subclass of the initializer() class with you particular behavior
+        self.initializer = Initializer(
+            description="initializer for feb 11 2022",
```
Comment (cdc-ap66) [inline src/dynode/config/covid_seip_config.py:27]: I'm still catching up on all the things you can do with pydantic, but it feels like you are both defining a class here and populating a config at the same time. Is this because covid has such a specific set of constraints that you would never be touching in experiments? Is this true will all initialization/configuration for this model config class?

also, you might want to call this class `SEIPCovidModel` just to be more specific

File: src/dynode/__init__.py [hunk 2]
```diff
@@ -35,6 +35,8 @@
     jax.Array,
 ]
 
+CompartmentGradiants = tuple[jax.Array]
```
[the remark this answers is not in the document]
Comment (arik-shurygin) [reply src/dynode/__init__.py:38]: Good point, I think that if the user calls `from dynode import *` they should get all the dynode modeling and the `config` module as well right.

File: src/dynode/config/dimension.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class Dimension(BaseModel):
+    """A dimension for a compartment"""
+
+    name: str
+    bins: Union[
+        List[CategoricalBin],
+        List[DiscretizedPositiveIntBin],
+        List[AgeBin],
```
[the remark this answers is not in the document]
Comment (arik-shurygin) [reply src/dynode/config/dimension.py:16]: If this typing is not strict could we instead create a base class `bin` and have all types of bins be children of that class? This way we dont need to constantly add things to this type definition

File: src/dynode/config/covid_seip_config.py (hunk 1 above)
Comment (arik-shurygin) [reply src/dynode/config/covid_seip_config.py:27]: This file was mostly to show how a user may specify a particular config, copying the values from an experiment run recently in production. This will likely not be included when we ship this PR, it will instead be ported to `DynODE-Experiments`

File: src/dynode/config/dimension.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cdc-ap66) [reply src/dynode/config/dimension.py:16]: that might work. I do think you want to make sure that you can't mix bin types in your list

File: src/dynode/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arik-shurygin) [reply src/dynode/__init__.py:38]: added another gh issue to move over these changes to the `__init__.py` file, since it will cause many small refactoring changes across the other files within `src/` and their imports and would bloat this PR: https://github.com/CDCgov/DynODE/issues/352

File: src/dynode/config/dimension.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arik-shurygin) [reply src/dynode/config/dimension.py]: thats a good point, maybe we add some sort of validator to check that all bins are of same type

File: src/dynode/pydantic_config/dimension.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    name: str
+    bins: List[Bin]
+
+    def __len__(self):
+        """Get len of a Dimension."""
+        return len(self.bins)
+
+    @model_validator(mode="after")
+    def check_bins_same_type(self) -> Self:
```
Comment (cdc-ap66) [inline src/dynode/pydantic_config/dimension.py:23]: very cool. Nice simplification

**Review by cdc-ap66 — APPROVED** [review]
LGTM!

**Comment by arik-shurygin** [comment]
I am curious for feedback from @kokbent  and @tjhladish  before I merge this in so I will keep it as a draft for right now.

**Comment by arik-shurygin** [comment]
A problem that arose in my mind is how exactly users will be expected to modify compartment values, since that is likely the most direct changes they will be making it should be easy and somewhat self explanatory.

I decided to put together 3 approaches, each with their own upsides and downsides to discuss:

we reference a `model.s` call in the following example, just know that does is call 
```
@property
    def s(self) -> Compartment:
        return self.compartments[0]
```
Here are the three approaches: 
```
from dynode.pydantic_config.covid_seip_config import SEIPCovidModel
model = SEIPCovidModel()
# testing out different methods of interacting with compartments
# the major issue is that jax arrays are immutable, meaning we must always have a reference to the Compartment's `values` property somewhere
# we can do it explicitly as in the first example, or hide it behind a wrapper like in the second.

# the most explicit form
model.compartments[0].values = (
    model.compartments[0].values.at[0, :, :, :].set(1.0)
)

# using `.s` property to return the Compartment object without indexing the `compartments` list 
model.s.values = model.s.values.at[0, :, :, :].set(1.0)

# using an overriden __setitem__() to approach a more numpy method of setting
# this is fighting against jax's functional nature and may bring upon problems in the future (?)
# on of the benefits of this is we can pass `Dimension` objects into the slices as well to make it very clear what we are modifying
model.s[1, :, :, :] = 2.0
print(model.compartments[0])
```

I think the first, most explicit example, is easy to get rid of. It is full of magic numbers and is hard to understand.

The second is a bit better, because we now have a `.s` property to identify the compartment but still, identifying the dimensions via integers is hard for newer modelers to dynode to grasp. 

I think that the last, numpy approach, is easiest for people to pick up on, but also goes against the fundamental nature of jax, which is to be a functional library. I am unsure what sorts of problems may arise by breaking this paradigm, and what sorts of things may become much more difficult to do as a result.

File: src/dynode/pydantic_config/config_definition.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        target_values_shape: tuple[int, ...] = tuple(
+            [len(d_i) for d_i in self.dimensions]
+        )
+        if self.values is not None:
+            assert target_values_shape == self.values.shape
+        else:
+            # fill with default for now, values filled in at runtime.
+            self.values = jnp.zeros(target_values_shape)
+        return self
+
```
Comment (cdc-ap66) [inline src/dynode/pydantic_config/config_definition.py:47]: should we consider initializing `.values` with an empty array if none is given. then you don't have to add assertion in the `__setitem__`.

File: src/dynode/pydantic_config/config_definition.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+from .dimension import Dimension
+from .strains import Strain
+
+
+class Compartment(BaseModel):
+    """Defines a single compartment of an ODE model."""
+
+    # allow jax array objects within Compartments
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    name: str
```
Comment (cdc-ap66) [inline src/dynode/pydantic_config/config_definition.py:31]: I think we should make sure this has some regex validation. no caps, no spaces, no special chars, only lowers and underscores

File: src/dynode/pydantic_config/config_definition.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    inference_method: Optional[MCMC | SVI] = None
+
+    def get_compartment(self, compartment_name: str):
+        for compartment in self.compartments:
+            if compartment_name == compartment.name:
+                return compartment
+        raise AssertionError(
+            "Compartment with name %s not found in model, found only these names: %s"
+            % (compartment_name, str([c.name for c in self.compartments]))
+        )
```
Comment (cdc-ap66) [inline src/dynode/pydantic_config/config_definition.py:127]: I think there might be a way to make things pseudo hashable, maybe with frozen config or perhaps by having setitem and getitem in component, then we can get this for free: https://docs.pydantic.dev/latest/concepts/models/#fields-with-non-hashable-default-values

File: src/dynode/pydantic_config/config_definition.py (hunk 5 above)
Comment (cdc-ap66) [reply src/dynode/pydantic_config/config_definition.py:47]: maybe something like this before the validation? This would also simplify the validation, only checking the shape if a jax array was provided

```python
    def model_post_init(self, *args, **kwargs): # I think this is the method we should use in pydantic
        if self.values is None:
            self.values = jnp.zeros(self.shape)
        self.__setattr__('at', self.values.at)
```

File: src/dynode/pydantic_config/config_definition.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    initializer: Initializer
+    compartments: List[Compartment]
+    parameters: ParamStore
+    # passed to diffrax.diffeqsolve
+    ode_function: Callable[
+        [List[Compartment], PositiveFloat, ParamStore], CompartmentGradiants
+    ]
+    # includes observation method, specified at runtime.
+    inference_method: Optional[MCMC | SVI] = None
+
```
Comment (cdc-ap66) [inline src/dynode/pydantic_config/config_definition.py:119]: how about we add something like this?:

```python
    def model_post_init(self, *args, **kwargs):
        for ci in self.compartments:
            self.__setattr__(ci.name, ci) # since compartments will have .at and be indexable as jax arrays
```

File: src/dynode/pydantic_config/config_definition.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        )
+
+
+class CompartmentalModel(BaseModel):
+    """An ODE compartment model configuration file."""
+
+    # allow users to pass custom types into CompartmentalModel
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    initializer: Initializer
+    compartments: List[Compartment]
```
Comment (cdc-ap66) [inline src/dynode/pydantic_config/config_definition.py:111]: I think there might be a way to make things pseudo hashable, maybe with frozen config or perhaps by having setitem and getitem in component, then we can get this for free: https://docs.pydantic.dev/latest/concepts/models/#fields-with-non-hashable-default-values

meaning, maybe the compartments nee simply to be indexable via .at or have both `__setitem__` and `__getitem__`

File: src/dynode/pydantic_config/config_definition.py (hunk 6 above)
Comment (cdc-ap66) [reply src/dynode/pydantic_config/config_definition.py:31]:
```python
    @field_validator('name', mode='before')
    @classmethod
    def make_attr_compliant(cls, value: str) -> str:
        if value.replace('_', '').isalpha():
            return value.lower()
        else:
            raise ValueError(
                "the name field must not contain non-alpha chars with the exception of underscores"
            )
```

File: src/dynode/pydantic_config/config_definition.py (hunk 5 above)
Comment (arik-shurygin) [reply src/dynode/pydantic_config/config_definition.py:47]: trying to figure out how this works, getting all sorts of weird errors because technically `.at` is not a field within the pydantic class? Sorting it out

File: src/dynode/model_configuration/config_definition.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        Compartment
+            Compartment class with matching name.
+
+        Raises
+        ------
+        AssertionError
+            raise if `compartment_name` not found within `self.compartments`
+        """
+        for compartment in self.compartments:
+            if compartment_name == compartment.name:
```
Comment (cdc-ap66) [inline src/dynode/model_configuration/config_definition.py:295]: I think we can eventually create a hidden name lookup at time of init so we don't need to have this conditional statement, but it isn;t like we are cycling through that many compartments.

```python

self._compartment_name_to_index_lookup = {ci.name: idx for idx, ci in enumerate(compartments)}
...
try:
    return self.compatments[self._compartment_name_to_index_lookup[compartment_name]]
except KeyError:
    raise KeyError(f"'{compartment_name}' does not map to an existing compartment")
```

**Review by cdc-ap66 — APPROVED** [review]
I like this a lot. nice commenting and docstrings, helpful patterns for naming, LGTM
